File: sdks/typescript/src/otel.ts
import { createHash } from "node:crypto";

export interface SpanLike {
  attributes?: Record<string, unknown>;
  _attributes?: Record<string, unknown>;
  [key: string]: unknown;
}

export interface SpanProcessorLike {
  onStart?(span: SpanLike, parentContext?: unknown): void;
  onEnd?(span: SpanLike): void;
  shutdown?(): Promise<void> | void;
  forceFlush?(): Promise<void> | void;
}

const SECRET_PATTERNS: Array<[RegExp, string]> = [
  [/sk-ant-[a-zA-Z0-9_-]{20,}/g, "ANTHROPIC_KEY"],
  [/sk-[a-zA-Z0-9_-]{20,}/g, "OPENAI_KEY"],
  [/AIza[0-9A-Za-z\-_]{35}/g, "GEMINI_KEY"],
  [/ghp_[a-zA-Z0-9]{36}/g, "GITHUB_TOKEN"],
  [/AKIA[0-9A-Z]{16}/g, "AWS_ACCESS_KEY"],
  [/-----BEGIN [A-Z ]*PRIVATE KEY-----/g, "PRIVATE_KEY"],
  [/postgres(?:ql)?:\/\/[^\s"']+/g, "DATABASE_URL"],
];

const SIG_PATTERNS: RegExp[] = [
  /"signature"\s*:\s*"([^"]+)"/g,
  /"encrypted_content"\s*:\s*"([^"]+)"/g,
  /"thought_signature"\s*:\s*"([^"]+)"/g,
];

export function fingerprint(text: string): string {
  return createHash("sha256").update(text).digest("hex").slice(0, 8);
}

export function scrubText(text: string): string {
  let result = text;

  // 1. Scrub credentials
  for (const [pattern, kind] of SECRET_PATTERNS) {
    pattern.lastIndex = 0;
    result = result.replace(pattern, (match) => {
      const fp = fingerprint(match);
      return `[REDACTED:${kind}:${fp}]`;
    });
  }

  // 2. Scrub reasoning signatures
  for (const pattern of SIG_PATTERNS) {
    pattern.lastIndex = 0;
    result = result.replace(pattern, (match, rawSig: string) => {
      if (!rawSig.startsWith("sgh_") && !rawSig.startsWith("sg_env_")) {
        const handle = `sgh_inproc_${fingerprint(rawSig)}`;
        const key = match.split(":")[0];
        return `${key}: "${handle}"`;
      }
      return match;
    });
  }

  return result;
}

export function scrubObject<T>(obj: T): T {
  if (typeof obj === "string") {
    return scrubText(obj) as unknown as T;
  }
  if (Array.isArray(obj)) {
    return obj.map((item) => scrubObject(item)) as unknown as T;
  }
  if (obj !== null && typeof obj === "object") {
    const result: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(obj as Record<string, unknown>)) {
      if (
        (key === "signature" || key === "encrypted_content" || key === "thought_signature") &&
        typeof value === "string"
      ) {
        if (!value.startsWith("sgh_") && !value.startsWith("sg_env_")) {
          result[key] = `sgh_inproc_${fingerprint(value)}`;
          continue;
        }
      }
      result[key] = scrubObject(value);
    }
    return result as unknown as T;
  }
  return obj;
}

export class StateGuardSanitizingSpanProcessor implements SpanProcessorLike {
  private nextProcessor?: SpanProcessorLike;

  constructor(nextProcessor?: SpanProcessorLike) {
    this.nextProcessor = nextProcessor;
  }

  public onStart(span: SpanLike, parentContext?: unknown): void {
    if (this.nextProcessor?.onStart) {
      this.nextProcessor.onStart(span, parentContext);
    }
  }

  public onEnd(span: SpanLike): void {
    const attributes = span.attributes || span._attributes;
    if (attributes && typeof attributes === "object") {
      for (const [key, val] of Object.entries(attributes)) {
        if (typeof val === "string") {
          attributes[key] = scrubText(val);
        } else if (val !== null && typeof val === "object") {
          attributes[key] = scrubObject(val);
        }
      }
    }

    if (this.nextProcessor?.onEnd) {
      this.nextProcessor.onEnd(span);
    }
  }

  public async shutdown(): Promise<void> {
    if (this.nextProcessor?.shutdown) {
      await this.nextProcessor.shutdown();
    }
  }

  public async forceFlush(): Promise<void> {
    if (this.nextProcessor?.forceFlush) {
      await this.nextProcessor.forceFlush();
    }
  }
}

File: sdks/python/stateguard/otel.py
import re
import hashlib
import json
from typing import Any, Dict, Optional, Union

try:
    from opentelemetry.sdk.trace import SpanProcessor
    from opentelemetry.trace import Span
except ImportError:
    class SpanProcessor:  # type: ignore
        """Fallback base class when opentelemetry-sdk is not installed."""
        def on_start(self, span: Any, parent_context: Optional[Any] = None) -> None:
            pass
        def on_end(self, span: Any) -> None:
            pass
        def shutdown(self) -> None:
            pass
        def force_flush(self, timeout_millis: int = 30000) -> bool:
            return True

    class Span:  # type: ignore
        pass

SECRET_PATTERNS = [
    (re.compile(r"sk-ant-[a-zA-Z0-9_\-]{20,}"), "ANTHROPIC_KEY"),
    (re.compile(r"sk-[a-zA-Z0-9_\-]{20,}"), "OPENAI_KEY"),
    (re.compile(r"AIza[0-9A-Za-z\-_]{35}"), "GEMINI_KEY"),
    (re.compile(r"ghp_[a-zA-Z0-9]{36}"), "GITHUB_TOKEN"),
    (re.compile(r"AKIA[0-9A-Z]{16}"), "AWS_ACCESS_KEY"),
    (re.compile(r"-----BEGIN [A-Z ]*PRIVATE KEY-----"), "PRIVATE_KEY"),
    (re.compile(r"postgres(?:ql)?://[^\s\"']+"), "DATABASE_URL"),
]

SIG_PATTERNS = [
    re.compile(r'"signature"\s*:\s*"([^"]+)"'),
    re.compile(r'"encrypted_content"\s*:\s*"([^"]+)"'),
    re.compile(r'"thought_signature"\s*:\s*"([^"]+)"'),
]


def fingerprint(text: str) -> str:
    return hashlib.sha256(text.encode("utf-8")).hexdigest()[:8]


def scrub_text(text: str) -> str:
    """Sanitizes text by masking credentials and replacing raw reasoning envelopes."""
    result = text
    # 1. Scrub credentials
    for pattern, kind in SECRET_PATTERNS:
        matches = list(pattern.finditer(result))
        for m in reversed(matches):
            fp = fingerprint(m.group(0))
            mask = f"[REDACTED:{kind}:{fp}]"
            result = result[:m.start()] + mask + result[m.end():]

    # 2. Scrub raw signatures
    for pattern in SIG_PATTERNS:
        matches = list(pattern.finditer(result))
        for m in reversed(matches):
            raw_sig = m.group(1)
            if not raw_sig.startswith("sgh_") and not raw_sig.startswith("sg_env_"):
                handle = f"sgh_inproc_{fingerprint(raw_sig)}"
                prefix = m.group(0).split(":")[0]
                result = result[:m.start()] + f'{prefix}: "{handle}"' + result[m.end():]

    return result


def scrub_object(obj: Any) -> Any:
    """Recursively scrub strings in dicts, lists, and primitives."""
    if isinstance(obj, str):
        return scrub_text(obj)
    elif isinstance(obj, dict):
        new_dict = {}
        for k, v in obj.items():
            if k in ("signature", "encrypted_content", "thought_signature") and isinstance(v, str):
                if not v.startswith("sgh_") and not v.startswith("sg_env_"):
                    new_dict[k] = f"sgh_inproc_{fingerprint(v)}"
                    continue
            new_dict[k] = scrub_object(v)
        return new_dict
    elif isinstance(obj, list):
        return [scrub_object(item) for item in obj]
    return obj


class StateGuardSanitizingSpanProcessor(SpanProcessor):
    """
    OpenTelemetry SpanProcessor that intercepts span attributes (gen_ai.prompt,
    gen_ai.completion, llm.output, etc.) and mutates them in-place,
    stripping credentials and raw reasoning envelopes before export.
    """

    TARGET_ATTRIBUTES = {
        "gen_ai.prompt",
        "gen_ai.completion",
        "llm.output",
        "llm.input",
        "gen_ai.request.headers",
        "gen_ai.response.headers",
        "traceloop.entity.input",
        "traceloop.entity.output",
    }

    def __init__(self, next_processor: Optional[SpanProcessor] = None):
        self.next_processor = next_processor

    def on_start(self, span: Any, parent_context: Optional[Any] = None) -> None:
        if self.next_processor:
            self.next_processor.on_start(span, parent_context)

    def on_end(self, span: Any) -> None:
        """Inspect and scrub span attributes in-memory."""
        attributes = getattr(span, "attributes", None)
        if attributes is None:
            attributes = getattr(span, "_attributes", None)

        if isinstance(attributes, dict):
            for key, val in list(attributes.items()):
                if isinstance(val, str):
                    attributes[key] = scrub_text(val)
                elif isinstance(val, (dict, list)):
                    attributes[key] = scrub_object(val)

        if self.next_processor:
            self.next_processor.on_end(span)

    def shutdown(self) -> None:
        if self.next_processor:
            self.next_processor.shutdown()

    def force_flush(self, timeout_millis: int = 30000) -> bool:
        if self.next_processor:
            return self.next_processor.force_flush(timeout_millis)
        return True
